### PR TITLE
feat: use post_commit hook to broadcast rlog

### PR DESCRIPTION
**Pull request by thalesmg**

### CODE REVIEW

File: src/mria_config.erl [hunk 1]
```diff
@@ -205,6 +209,10 @@ make_shard_match_spec(Tables) ->
      , ['$_']
      } || Table <- Tables].
 
+-spec otp_is_compatible() -> boolean().
+otp_is_compatible() ->
+    erlang:function_exported(mnesia_hook, register_hook, 2).
```
Comment (k32) [inline src/mria_config.erl:214]: This can be replaced with something like 
```
try mnesia_hook:module_info() 
of _ -> true 
catch error:undef -> false
end.
```
Then this can be removed: https://github.com/emqx/mria/pull/53/files#diff-29ae83af7897f38e4d8af84d372b03c8fee37ec5cdff612f5fd9e5941d7f4e00R31

File: src/mria_config.erl [hunk 2]
```diff
@@ -205,6 +209,17 @@ make_shard_match_spec(Tables) ->
      , ['$_']
      } || Table <- Tables].
 
+-spec otp_is_compatible() -> boolean().
+otp_is_compatible() ->
+    %% for some reason, `erlang:function_exported' does not see
+    %% `mnesia_hook:register_hook/2' even after loading the mnesia
+    %% application.  So this is needed to make it "visible".
```
Comment (k32) [inline src/mria_config.erl:216]: Nit: this comment is no longer relevant, since we can pretend that the previous implementation never happened.

File: src/mria_lib.erl [hunk 3]
```diff
@@ -30,8 +29,6 @@
         , cancel_timer/1
         , subscriber_node/1
 
-        , get_internals/0
```
Comment (k32) [inline src/mria_lib.erl:33]: Nit: this function is useful for debugging, maybe leave it

File: src/mria_rlog.erl [hunk 4]
```diff
@@ -124,3 +138,38 @@ subscribe(Shard, RemoteNode, Subscriber, Checkpoint) ->
...
+%% TODO: Handle local content tables more soundly.
+detect_shard(#{ram_copies := [Op | _]}) ->
+    do_detect_shard(Op);
+detect_shard(#{disc_copies := [Op | _]}) ->
+    do_detect_shard(Op);
+detect_shard(#{disc_only_copies := [Op | _]}) ->
+    do_detect_shard(Op);
+detect_shard(#{ext := [{ext_copies, [{_, Op}]} | _]}) ->
+    do_detect_shard(Op);
+%% FIXME: Should we handle snmp ext type ops? how to test?
```
Comment (k32) [inline src/mria_rlog.erl:159]: Let's assume that we don't support SNMP.

File: src/mria_rlog_tab.erl [hunk 5]
```diff
@@ -21,8 +21,6 @@
 %% Mnesia bootstrap
 -export([ensure_table/1]).
 
--export([write/3]).
```
Comment (k32) [inline src/mria_rlog_tab.erl:24]: The comment for the module is no longer relevant:
```
%% access module for transaction logs
%% implemented as a ram_copies mnesia table
```

Comment (k32) [reply src/mria_rlog_tab.erl:24]: Also this module itself is a bit useless. `ensure_table` can be moved to `mria_lib`, for example.

File: src/mria_status.erl [hunk 6]
```diff
@@ -204,10 +203,6 @@ get_shard_stats(Shard) ->
 notify_replicant_state(Shard, State) ->
     set_stat(Shard, ?replicant_state, State).
 
--spec notify_replicant_import_trans(mria_rlog:shard(), mria_rlog_server:checkpoint()) -> ok.
```
Comment (k32) [inline src/mria_status.erl:207]: The initial idea was to keep a metric that can be useful to estimate replicants' lag in comparison with the core. I think this function can still be useful, just use wall clock instead of checkpoint. But this can be left for another PR, just keep the function with a TODO, perhaps.

File: src/mria_rlog_tab.erl (hunk 5 above)
Comment (thalesmg) [reply src/mria_rlog_tab.erl:24]: I tried to delete it at first, but a few synchronization things depend on waiting on this table.

I thought that we would perhaps remove this table later as it might touch other things, but I'll try to delete and change the places that wait on the shard and see if it's too big.

Comment (k32) [reply src/mria_rlog_tab.erl:24]: Ok. This cleanup can be left for the next PR.

File: test/mria_SUITE.erl [hunk 7]
```diff
@@ -625,6 +625,57 @@ t_rlog_schema(_) ->
                             ))
        end).
 
+%% Test post commit hook using disc tables.
```
Comment (k32) [inline test/mria_SUITE.erl:628]: I would like to see a smoke test that verifies the assumption that post_commit hook rans on all core nodes in the cluster for both transactions and the dirty operations.

File: src/mria_rlog_agent.erl [hunk 8]
```diff
@@ -136,26 +136,27 @@ handle_state_trans(_OldState, _State, _Data) ->
...
-handle_mnesia_event({Shard, TXID, Ops}, _ActivityId, D = #d{shard = Shard}) ->
+-spec handle_mnesia_event({mria_rlog:shard(), mria_rlog:commit_records()}, term(), data()) ->
+          fsm_result().
+handle_mnesia_event({Shard, Commit}, ActivityId, D = #d{shard = Shard}) ->
     PushMode = D#d.push_mode,
     SeqNo    = D#d.seqno,
+    Ops = maps:fold(
+            fun(K, Ops, Acc) when K =:= ram_copies;
+                                  K =:= disc_copies;
+                                  K =:= disc_only_copies ->
```
Comment (k32) [inline src/mria_rlog_agent.erl:147]: Ext copies are missing?

File: src/mria_rlog_replica.erl [hunk 9]
```diff
@@ -156,26 +156,27 @@ push_tlog_entry(sync, Shard, {Node, Pid}, Batch) ->
...
+         , transaction      => Transaction
+         , tid              => _Tid
          }),
-    Checkpoint = mria_lib:txid_to_checkpoint(TXID),
     ok = mria_replica_importer_worker:import_batch(ImporterWorker, Transaction),
-    mria_status:notify_replicant_import_trans(Shard, Checkpoint),
+    %% statistic to give an estimate the replicant lag with respect to
+    %% the core node.
+    Timestamp = erlang:timestamp(),
+    mria_status:notify_replicant_import_trans(Shard, Timestamp),
```
Comment (k32) [inline src/mria_rlog_replica.erl:176]: The timestamp should be collected on the core node?

File: src/mria_rlog_replica.erl [hunk 10]
```diff
@@ -185,18 +186,13 @@ handle_tlog_entry(St, {Agent, SeqNo, TXID, Transaction},
     ?tp(rlog_replica_store_trans,
         #{ agent       => Agent
          , seqno       => SeqNo
-         , txid        => TXID
          , transaction => Transaction
+         , tid              => _Tid
```
Comment (k32) [inline src/mria_rlog_replica.erl:190]: Nit: indentation

File: test/mria_SUITE.erl (hunk 7 above)
Comment (thalesmg) [reply test/mria_SUITE.erl]: I've changed this test to check the events of event being intercepted, agent receiving the event and its import.
Anything else of interest that we can check?

File: test/mria_rlog_props.erl [hunk 11]
```diff
@@ -56,6 +57,42 @@ replicant_bootstrap_stages(Node, Trace0) ->
...
+              || Event = #{ ram_copies := RamCopies
+                          , schema_ops := []
+                          } <- Trace0,
+                 case RamCopies of
+                     [{{mria_schema, _}, _, _} | _] -> false;
+                     _ -> true
+                 end
+            ],
+    ?assert(
+       ?causality(
```
[the remark this answers is not in the document]
Comment (thalesmg) [reply test/mria_rlog_props.erl:72]: There's one case that might have a "cause" without an effect here.

Suppose there are 2 cores, 2 replicants.  One of the cores ("Core A") holds the 2 agents for the 2 replicants, the other core ("Core B") has no agents.  In that case, `mria_rlog_intercept_trans` from the core with no agents will have no matching `rlog_realtime_op` effect.  Some time it'll pass (by luck), but in others it might happen the following:

1. `mria_rlog_intercept_trans` in Core A
2. `rlog_realtime_op` in Agent 1 (node of Core A)
3. `rlog_realtime_op` in Agent 2 (node of Core A)
4. `mria_rlog_intercept_trans` in Core B (no agent to pass this on)

One simple way to avoid this in the test is to use a single core.  I'm still thinking of a way to encode the property without changing the number of cores.

[the remark this answers is not in the document]
Comment (k32) [reply test/mria_rlog_props.erl:72]: I see, thanks. Something like this will help?
```
[?strict_causality(#{?snk_kind := mria_rlog_intercept_trans, tid := _Tid}, 
                   #{?snk_kind := rlog_realtime_op, ?snk_meta := #{pid := AgentPid}, activity_id := _Tid},
                   Trace) || AgentPid <- AgentPids] %% pids are aquired by scanning the trace
).
```

File: src/mria_lib.erl [hunk 12]
```diff
@@ -85,10 +84,11 @@
...
 -type rlog() :: #rlog{}.
 
--type tlog_entry() :: { _Sender :: pid()
-                      , _SeqNo  :: integer()
-                      , _Key    :: txid()
-                      , _Tx     :: [tx()]
+-type tlog_entry() :: { _Sender    :: pid()
+                      , _SeqNo     :: integer()
+                      , _Tid       :: {tid, term(), term()} | {dirty, term()}
+                      , _Timestamp :: erlang:timestamp()
```
Comment (k32) [inline src/mria_lib.erl:90]: Currently each transaction record contains both seqno and timestamp.
They can be unified: currently seqno is generated by the agent, so it restarts from 0 on every connection
What if instead SeqNo is generated by mria_server and is transferred in the handshake, so it approaches the same value on all the replicant nodes connected to the given core node. Then lag can be calculated correctly when there are no transactions in the shard for a long time.
Also timestamp won't be needed anymore (save some bandwidth)

File: test/mria_rlog_props.erl [hunk 13]
```diff
@@ -56,6 +57,66 @@ replicant_bootstrap_stages(Node, Trace0) ->
...
+                 end
+            ],
+    ?assert(
+       ?strict_causality(
+          #{ ?snk_kind    := mria_rlog_intercept_trans
+           , tid          := _Tid
+           } = _CommitRecord
+         , #{ ?snk_kind   := rlog_realtime_op
+            , activity_id := _Tid
+              , ops         := _ImportOps
```
Comment (k32) [inline test/mria_rlog_props.erl:89]: Nit: indentation

File: test/mria_rlog_props.erl [hunk 14]
```diff
@@ -56,6 +57,66 @@ replicant_bootstrap_stages(Node, Trace0) ->
...
+            }
+          , length(ops_from_commit_record(_CommitRecord)) =:= length(_ImportOps)
+         , Trace
+         )
+      ),
```
Comment (k32) [inline test/mria_rlog_props.erl:94]: Nit, coding style: in the rest of the code closing parenthesis are usually glued together lisp-style to save space.

File: src/mria_rlog_replica.erl [hunk 15]
```diff
@@ -349,6 +343,11 @@ try_connect([Node|Rest], Shard, Checkpoint) ->
          }),
     case mria_rlog:subscribe(Shard, Node, self(), Checkpoint) of
         {ok, NeedBootstrap, Agent, TableSpecs} ->
+            ?tp(notice, "Connected to the core node",
+                #{ shard => Shard
+                 , node  => Node
+                 , agent => self()
```
Comment (k32) [inline src/mria_rlog_replica.erl:349]: Use of `agent` term here is misleading, since it usually refers to `mria_agent`. Also pid can be found in the event metadata, so it's not strictly necessary here.

File: test/mria_rlog_props.erl [hunk 16]
```diff
@@ -56,6 +57,63 @@ replicant_bootstrap_stages(Node, Trace0) ->
...
               ).
 
+%% Check that all commit logs intercepted are received by an agent
+all_intercepted_commit_logs_received(Trace0) ->
+    %% we consider only core nodes that have following agents, as only
+    %% those have a matching effect for the intercept event.
+    NodesWithAgents =
+        lists:usort([Node || #{node := Node} <- ?of_kind("Connected to the core node", Trace0)]),
+    Trace = [ Event
+              || Event <- ?of_kind([ mria_rlog_intercept_trans
```
Comment (k32) [inline test/mria_rlog_props.erl:67]: Nit: it's more efficient to match kind in the same list comprehension rather than invoke `?of_kind`, this way filtering will be done in one pass rather than two.

File: test/mria_SUITE.erl [hunk 17]
```diff
@@ -625,6 +625,95 @@ t_rlog_schema(_) ->
...
+                                         , {record_name, kv_tab}
+                                         , {attributes, record_info(fields, kv_tab)}
+                                         ]
+                                       ])
+             end,
+             [ {kv_tab1, disc_copies}
+             , {kv_tab2, disc_only_copies}
+             , {kv_tab3, ram_copies}
+             , {kv_tab4, rocksdb_copies}
+             ]),
```
Comment (k32) [inline test/mria_SUITE.erl:651]: Can be factored to a function (smth. like `create_test_tables([{kv_tab1, disc_copies}, ...])`) to simplify reading the testcase?

File: test/mria_SUITE.erl [hunk 18]
```diff
@@ -625,6 +625,95 @@ t_rlog_schema(_) ->
...
+                                           mnesia:write(kv_tab4, {kv_tab, w4, w4}, write),
+                                           ok
+                                   end]),
+           ok = rpc:call(N3, mria, dirty_write, [kv_tab1, {kv_tab, dw1, dw1}]),
+           ok = rpc:call(N3, mria, dirty_write, [kv_tab2, {kv_tab, dw2, dw2}]),
+           ok = rpc:call(N3, mria, dirty_write, [kv_tab3, {kv_tab, dw3, dw3}]),
+           ok = rpc:call(N3, mria, dirty_write, [kv_tab4, {kv_tab, dw4, dw4}]),
+           mria_mnesia_test_util:wait_full_replication(Cluster),
+           %% other replica should get updates
+           {atomic, Res} = rpc:call(N4, mria, transaction,
```
Comment (k32) [inline test/mria_SUITE.erl:669]: Can be factored to a helper function (e.g. `compare_shard_contents(CoreNode, ReplicantNode)`)

**Review by k32 — APPROVED** [review]

**Comment by k32** [comment]
Looks good, the rest can be ironed out in the later PRs.

File: src/mria_rlog_tab.erl (hunk 5 above)
Comment (thalesmg) [reply src/mria_rlog_tab.erl:24]: You were right, removing this was simple.  I must have gotten confused while trying to remove it and thought that it'd require a lot of refactoring.  :speak_no_evil: :see_no_evil:
